File: server/internal/game/board.go
package game

import "sort"

type NodeType string

const (
	Station      NodeType = "station"
	Camp         NodeType = "camp"
	Headquarters NodeType = "headquarters"
)

type Node struct {
	ID        string   `json:"id"`
	X         float64  `json:"x"`
	Y         float64  `json:"y"`
	Type      NodeType `json:"type"`
	DeployFor Seat     `json:"deployFor,omitempty"`
	Row       int      `json:"row,omitempty"`
	Column    string   `json:"column,omitempty"`
}

type RailHeading string

const (
	RailNorth RailHeading = "north"
	RailEast  RailHeading = "east"
	RailSouth RailHeading = "south"
	RailWest  RailHeading = "west"
)

type Edge struct {
	From            string      `json:"from"`
	To              string      `json:"to"`
	Type            string      `json:"type"`
	RailwayOrigin   RailHeading `json:"railwayOrigin,omitempty"`
	RailwayTerminal RailHeading `json:"railwayTerminal,omitempty"`
}

type BoardDefinition struct {
	Version string          `json:"version"`
	Width   int             `json:"width"`
	Height  int             `json:"height"`
	Nodes   map[string]Node `json:"nodes"`
	Edges   []Edge          `json:"edges"`
}

type Board struct {
	Version string            `json:"version"`
	Width   int               `json:"width"`
	Height  int               `json:"height"`
	Nodes   map[string]Node   `json:"nodes"`
	Edges   []Edge            `json:"edges"`
	Adj     map[string][]Edge `json:"-"`
}

var localColumns = []string{"1L", "2L", "3", "2R", "1R"}

type localPosition struct {
	row int
	col int
}

// NewBoard creates the standard four-country Army Chess topology. Coordinates
// are drawing coordinates in a 1000x1000 view box; movement is determined only
// by the explicit graph below.
func NewBoard() *Board {
	b := &Board{
		Version: "board.v2",
		Width:   1000,
		Height:  1000,
		Nodes:   map[string]Node{},
		Adj:     map[string][]Edge{},
	}

	for _, seat := range Seats {
		for row := 1; row <= 6; row++ {
			for col := range localColumns {
				x, y := localCoordinates(row, col)
				x, y = rotateCoordinates(seat, x, y)
				nodeType := Station
				switch {
				case (row == 2 || row == 4) && (col == 1 || col == 3), row == 3 && col == 2:
					nodeType = Camp
				case row == 6 && (col == 1 || col == 3):
					nodeType = Headquarters
				}
				id := localNodeID(seat, row, col)
				b.Nodes[id] = Node{ID: id, X: x, Y: y, Type: nodeType, DeployFor: seat, Row: row, Column: localColumns[col]}
			}
		}
	}

	palace := map[string][2]float64{
		"palace-nw": {410, 410}, "palace-n": {500, 410}, "palace-ne": {590, 410},
		"palace-w": {410, 500}, "palace-c": {500, 500}, "palace-e": {590, 500},
		"palace-sw": {410, 590}, "palace-s": {500, 590}, "palace-se": {590, 590},
	}
	for id, point := range palace {
		b.Nodes[id] = Node{ID: id, X: point[0], Y: point[1], Type: Station}
	}

	for _, seat := range Seats {
		for _, row := range []int{2, 3, 4, 6} {
			for col := 0; col < 4; col++ {
				b.addRoad(localNodeID(seat, row, col), localNodeID(seat, row, col+1))
			}
		}
		for _, col := range []int{1, 2, 3} {
			for row := 2; row < 6; row++ {
				b.addRoad(localNodeID(seat, row, col), localNodeID(seat, row+1, col))
			}
		}
		for _, col := range []int{0, 4} {
			b.addRoad(localNodeID(seat, 5, col), localNodeID(seat, 6, col))
		}
		for _, row := range []int{1, 5} {
			for col := 0; col < 4; col++ {
				b.addRailBetween(localNodeID(seat, row, col), localNodeID(seat, row, col+1))
			}
		}
		for _, col := range []int{0, 4} {
			for row := 1; row < 5; row++ {
				b.addRailBetween(localNodeID(seat, row, col), localNodeID(seat, row+1, col))
			}
		}

		// The diagonal cross around every camp is a road junction. This includes
		// the five camps themselves; a camp is only protected when occupied.
		for _, camp := range []localPosition{{2, 1}, {2, 3}, {3, 2}, {4, 1}, {4, 3}} {
			for _, diagonal := range []localPosition{
				{camp.row - 1, camp.col - 1}, {camp.row - 1, camp.col + 1},
				{camp.row + 1, camp.col - 1}, {camp.row + 1, camp.col + 1},
			} {
				b.addRoad(localNodeID(seat, camp.row, camp.col), localNodeID(seat, diagonal.row, diagonal.col))
			}
		}
	}

	// The nine-palace is a 3x3 railway graph.
	palaceRows := [][]string{{"palace-nw", "palace-n", "palace-ne"}, {"palace-w", "palace-c", "palace-e"}, {"palace-sw", "palace-s", "palace-se"}}
	for row := range palaceRows {
		for col := 0; col < 2; col++ {
			b.addRailBetween(palaceRows[row][col], palaceRows[row][col+1])
		}
	}
	for col := range palaceRows[0] {
		for row := 0; row < 2; row++ {
			b.addRailBetween(palaceRows[row][col], palaceRows[row+1][col])
		}
	}

	interfaces := map[Seat]map[int]string{
		North: {0: "palace-ne", 2: "palace-n", 4: "palace-nw"},
		East:  {0: "palace-se", 2: "palace-e", 4: "palace-ne"},
		South: {0: "palace-sw", 2: "palace-s", 4: "palace-se"},
		West:  {0: "palace-nw", 2: "palace-w", 4: "palace-sw"},
	}
	for _, seat := range Seats {
		for col, palaceID := range interfaces[seat] {
			b.addRailBetween(localNodeID(seat, 1, col), palaceID)
		}
	}

	// These four bent routes connect neighboring front corners. Their headings
	// are retained so ordinary railway pieces cannot turn at the corner.
	b.addBentRail(localNodeID(South, 1, 4), localNodeID(East, 1, 0), RailNorth, RailEast)
	b.addBentRail(localNodeID(East, 1, 4), localNodeID(North, 1, 0), RailWest, RailNorth)
	b.addBentRail(localNodeID(North, 1, 4), localNodeID(West, 1, 0), RailSouth, RailWest)
	b.addBentRail(localNodeID(West, 1, 4), localNodeID(South, 1, 0), RailEast, RailSouth)

	return b
}

func localCoordinates(row, col int) (float64, float64) {
	return 406 + float64(col)*47, 648 + float64(row-1)*47
}

func rotateCoordinates(seat Seat, x, y float64) (float64, float64) {
	switch seat {
	case North:
		return 1000 - x, 1000 - y
	case East:
		return y, 1000 - x
	case West:
		return 1000 - y, x
	default:
		return x, y
	}
}

func localNodeID(seat Seat, row, col int) string {
	return string(seat) + "-r" + string(rune('0'+row)) + "-" + localColumns[col]
}

func (b *Board) addDirected(edge Edge) {
	for _, existing := range b.Adj[edge.From] {
		if existing.To == edge.To && existing.Type == edge.Type {
			return
		}
	}
	b.Edges = append(b.Edges, edge)
	b.Adj[edge.From] = append(b.Adj[edge.From], edge)
}

func (b *Board) addRoad(from, to string) {
	b.addDirected(Edge{From: from, To: to, Type: "road"})
	b.addDirected(Edge{From: to, To: from, Type: "road"})
}

func (b *Board) addRailBetween(from, to string) {
	fromNode, fromOK := b.Nodes[from]
	toNode, toOK := b.Nodes[to]
	if !fromOK || !toOK {
		return
	}
	heading := headingBetween(fromNode, toNode)
	b.addDirected(Edge{From: from, To: to, Type: "rail", RailwayOrigin: heading, RailwayTerminal: heading})
	b.addDirected(Edge{From: to, To: from, Type: "rail", RailwayOrigin: oppositeHeading(heading), RailwayTerminal: oppositeHeading(heading)})
}

func (b *Board) addBentRail(from, to string, first, second RailHeading) {
	b.addDirected(Edge{From: from, To: to, Type: "rail", RailwayOrigin: first, RailwayTerminal: second})
	b.addDirected(Edge{From: to, To: from, Type: "rail", RailwayOrigin: oppositeHeading(second), RailwayTerminal: oppositeHeading(first)})
}

func headingBetween(from, to Node) RailHeading {
	if from.X == to.X {
		if to.Y < from.Y {
			return RailNorth
		}
		return RailSouth
	}
	if to.X < from.X {
		return RailWest
	}
	return RailEast
}

func oppositeHeading(heading RailHeading) RailHeading {
	switch heading {
	case RailNorth:
		return RailSouth
	case RailEast:
		return RailWest
	case RailSouth:
		return RailNorth
	default:
		return RailEast
	}
}

func (b *Board) Definition() BoardDefinition {
	nodes := make(map[string]Node, len(b.Nodes))
	for id, node := range b.Nodes {
		nodes[id] = node
	}
	edges := append([]Edge(nil), b.Edges...)
	return BoardDefinition{Version: b.Version, Width: b.Width, Height: b.Height, Nodes: nodes, Edges: edges}
}

func (b *Board) DeploymentNodes(seat Seat) []string {
	ids := make([]string, 0, 25)
	for id, node := range b.Nodes {
		if node.DeployFor == seat && node.Type != Camp {
			ids = append(ids, id)
		}
	}
	sort.Strings(ids)
	return ids
}

func (b *Board) Node(id string) (Node, bool) { node, ok := b.Nodes[id]; return node, ok }

func (b *Board) DirectEdge(from, to string) (Edge, bool) {
	for _, edge := range b.Adj[from] {
		if edge.To == to {
			return edge, true
		}
	}
	return Edge{}, false
}

File: web/src/board.ts
export type BoardSeat = "north" | "east" | "south" | "west";
export type BoardNodeType = "station" | "camp" | "headquarters" | "frontline" | "mountain";
export type BoardEdgeType = "road" | "rail";

export type BoardNode = {
  id: string;
  x: number;
  y: number;
  type: BoardNodeType;
  deployFor?: BoardSeat;
  row?: number;
  column?: string;
};

export type BoardEdge = { from: string; to: string; type: BoardEdgeType };
export type BoardDefinition = { version: string; width: number; height: number; nodes: Record<string, BoardNode>; edges: BoardEdge[] };

const columns = ["1L", "2L", "3", "2R", "1R"];
const seats: BoardSeat[] = ["north", "east", "south", "west"];

function localID(seat: BoardSeat, row: number, col: number) {
  return `${seat}-r${row}-${columns[col]}`;
}

function rotate(seat: BoardSeat, x: number, y: number): [number, number] {
  if (seat === "north") return [1000 - x, 1000 - y];
  if (seat === "east") return [y, 1000 - x];
  if (seat === "west") return [1000 - y, x];
  return [x, y];
}

function buildBoard(): BoardDefinition {
  const nodes: Record<string, BoardNode> = {};
  const edges: BoardEdge[] = [];
  const edgeKeys = new Set<string>();
  const gridStep = 47;
  const formationStart = 406;
  const frontRow = 648;

  function addEdge(from: string, to: string, type: BoardEdgeType) {
    const key = [from, to].sort().join("|");
    if (edgeKeys.has(key)) return;
    edgeKeys.add(key);
    edges.push({ from, to, type });
  }

  for (const seat of seats) {
    for (let row = 1; row <= 6; row += 1) {
      for (let col = 0; col < columns.length; col += 1) {
        const [x, y] = rotate(seat, formationStart + col * gridStep, frontRow + (row - 1) * gridStep);
        const camp = (row === 2 || row === 4) && (col === 1 || col === 3) || row === 3 && col === 2;
        const headquarters = row === 6 && (col === 1 || col === 3);
        nodes[localID(seat, row, col)] = {
          id: localID(seat, row, col), x, y,
          type: camp ? "camp" : headquarters ? "headquarters" : "station",
          deployFor: seat, row, column: columns[col],
        };
      }
    }
  }

  const palace: Record<string, [number, number]> = {
    "palace-nw": [410, 410], "palace-n": [500, 410], "palace-ne": [590, 410],
    "palace-w": [410, 500], "palace-c": [500, 500], "palace-e": [590, 500],
    "palace-sw": [410, 590], "palace-s": [500, 590], "palace-se": [590, 590],
  };
  for (const [id, [x, y]] of Object.entries(palace)) nodes[id] = { id, x, y, type: "station" };

  for (const seat of seats) {
    for (const row of [2, 3, 4, 6]) for (let col = 0; col < 4; col += 1) addEdge(localID(seat, row, col), localID(seat, row, col + 1), "road");
    for (const col of [1, 2, 3]) for (let row = 2; row < 6; row += 1) addEdge(localID(seat, row, col), localID(seat, row + 1, col), "road");
    for (const col of [0, 4]) addEdge(localID(seat, 5, col), localID(seat, 6, col), "road");
    for (const row of [1, 5]) for (let col = 0; col < 4; col += 1) addEdge(localID(seat, row, col), localID(seat, row, col + 1), "rail");
    for (const col of [0, 4]) for (let row = 1; row < 5; row += 1) addEdge(localID(seat, row, col), localID(seat, row + 1, col), "rail");

    for (const [row, col] of [[2, 1], [2, 3], [3, 2], [4, 1], [4, 3]]) {
      for (const [deltaRow, deltaCol] of [[-1, -1], [-1, 1], [1, -1], [1, 1]]) {
        addEdge(localID(seat, row, col), localID(seat, row + deltaRow, col + deltaCol), "road");
      }
    }
  }

  const palaceRows = [["palace-nw", "palace-n", "palace-ne"], ["palace-w", "palace-c", "palace-e"], ["palace-sw", "palace-s", "palace-se"]];
  for (const row of palaceRows) for (let col = 0; col < 2; col += 1) addEdge(row[col], row[col + 1], "rail");
  for (let col = 0; col < 3; col += 1) for (let row = 0; row < 2; row += 1) addEdge(palaceRows[row][col], palaceRows[row + 1][col], "rail");

  const interfaces: Record<BoardSeat, Record<number, string>> = {
    north: { 0: "palace-ne", 2: "palace-n", 4: "palace-nw" },
    east: { 0: "palace-se", 2: "palace-e", 4: "palace-ne" },
    south: { 0: "palace-sw", 2: "palace-s", 4: "palace-se" },
    west: { 0: "palace-nw", 2: "palace-w", 4: "palace-sw" },
  };
  for (const seat of seats) for (const [col, palaceID] of Object.entries(interfaces[seat])) addEdge(localID(seat, 1, Number(col)), palaceID, "rail");
  addEdge(localID("south", 1, 4), localID("east", 1, 0), "rail");
  addEdge(localID("east", 1, 4), localID("north", 1, 0), "rail");
  addEdge(localID("north", 1, 4), localID("west", 1, 0), "rail");
  addEdge(localID("west", 1, 4), localID("south", 1, 0), "rail");

  return { version: "board.v2", width: 1000, height: 1000, nodes, edges };
}

export const boardDefinition = buildBoard();

function buildOneVsOneBoard(): BoardDefinition {
  const nodes: Record<string, BoardNode> = {};
  const edges: BoardEdge[] = [];
  const edgeKeys = new Set<string>();
  const columns = ["1L", "2L", "3", "2R", "1R"];

  function addEdge(from: string, to: string, type: BoardEdgeType) {
    const key = [from, to].sort().join("|");
    if (edgeKeys.has(key)) return;
    edgeKeys.add(key);
    edges.push({ from, to, type });
  }

  function addZone(seat: "north" | "south", startY: number) {
    for (let row = 1; row <= 6; row += 1) {
      for (let col = 0; col < columns.length; col += 1) {
        const id = `${seat}-r${row}-${columns[col]}`;
        const camp = (row === 2 || row === 4) && (col === 1 || col === 3) || row === 3 && col === 2;
        const headquarters = row === 6 && (col === 1 || col === 3);
        nodes[id] = {
          id, x: 100 + col * 140, y: startY + (row - 1) * 75,
          type: camp ? "camp" : headquarters ? "headquarters" : "station",
          deployFor: seat, row, column: columns[col],
        };
      }
    }

    for (const row of [2, 3, 4, 6]) for (let col = 0; col < 4; col += 1) addEdge(`${seat}-r${row}-${columns[col]}`, `${seat}-r${row}-${columns[col + 1]}`, "road");
    for (const col of [1, 2, 3]) for (let row = 2; row < 6; row += 1) addEdge(`${seat}-r${row}-${columns[col]}`, `${seat}-r${row + 1}-${columns[col]}`, "road");
    for (const col of [0, 4]) addEdge(`${seat}-r5-${columns[col]}`, `${seat}-r6-${columns[col]}`, "road");
    for (const row of [1, 5]) for (let col = 0; col < 4; col += 1) addEdge(`${seat}-r${row}-${columns[col]}`, `${seat}-r${row}-${columns[col + 1]}`, "rail");
    for (const col of [0, 4]) for (let row = 1; row < 5; row += 1) addEdge(`${seat}-r${row}-${columns[col]}`, `${seat}-r${row + 1}-${columns[col]}`, "rail");

    for (const [row, col] of [[2, 1], [2, 3], [3, 2], [4, 1], [4, 3]]) {
      for (const [deltaRow, deltaCol] of [[-1, -1], [-1, 1], [1, -1], [1, 1]]) {
        addEdge(`${seat}-r${row}-${columns[col]}`, `${seat}-r${row + deltaRow}-${columns[col + deltaCol]}`, "road");
      }
    }
  }

  addZone("north", 100);
  addZone("south", 785);
  for (const [id, x, type] of [
    ["frontline-1L", 100, "frontline"], ["mountain-2L", 240, "mountain"],
    ["frontline-3", 380, "frontline"], ["mountain-2R", 520, "mountain"],
    ["frontline-1R", 660, "frontline"],
  ] as const) nodes[id] = { id, x, y: 630, type };
  for (const [top, frontline, bottom] of [
    ["north-r1-1L", "frontline-1L", "south-r1-1L"],
    ["north-r1-3", "frontline-3", "south-r1-3"],
    ["north-r1-1R", "frontline-1R", "south-r1-1R"],
  ]) {
    addEdge(top, frontline, "rail");
    addEdge(frontline, bottom, "rail");
  }
  return { version: "board.1v1", width: 760, height: 1260, nodes, edges };
}

export const oneVsOneBoardDefinition = buildOneVsOneBoard();

export function boardForMatchMode(matchMode?: "two_vs_two" | "one_vs_one"): BoardDefinition {
  return matchMode === "one_vs_one" ? oneVsOneBoardDefinition : boardDefinition;
}
